### PR TITLE
TDD: add test for nested summary work to be done

### PR DESCRIPTION
**Pull request by raubitsj**
test driven development?

### CODE REVIEW

**Comment by vanpelt** [comment]
Love it 😍

**Comment by davidwallacejackson** [comment]
It's shorter than the one in cli-og now, and it passes the tests. Is it better/more readable...? 🤷 

The code reuse needs felt like they fit this use of `abc`, but there's something kind of unintuitive about the result: I struggle to explain why `_get_full_path()` is the "abstract part" of this class. Curious what you guys think.

**Comment by davidwallacejackson** [comment]
Per the test failures: there's a lot of places I could add the logic to coerce this to a dictionary when serializing JSON -- could we pick a "right way" to do that, or at least define which ways are correct under which contexts? For instance, I could:

* add support for the Summary object to `json_friendly`
* add a check for the `json_encode` method to `json_friendly` and then implement that method for the base class
* add support for Summary separately to each JSONEncoder that handles Summary objects during tests

I think the second one is the most generic -- and doesn't require `util` having knowledge of the Summary class' internals. But it doesn't look like that mechanism (implemented [here](https://github.com/wandb/client-ng/blob/080a5c6a073ad9230f445ca1e0920f065bd14dd7/wandb/util.py#L482)) is being used anywhere at the moment, while the commented-out `to_json()` method below is defined for a bunch of different classes in `data_types.py`.

**Comment by raubitsj** [comment]
> Per the test failures: there's a lot of places I could add the logic to coerce this to a dictionary when serializing JSON -- could we pick a "right way" to do that, or at least define which ways are correct under which contexts? For instance, I could:
> 
> * add support for the Summary object to `json_friendly`
> * add a check for the `json_encode` method to `json_friendly` and then implement that method for the base class
> * add support for Summary separately to each JSONEncoder that handles Summary objects during tests
> 
> I think the second one is the most generic -- and doesn't require `util` having knowledge of the Summary class' internals. But it doesn't look like that mechanism (implemented [here](https://github.com/wandb/client-ng/blob/080a5c6a073ad9230f445ca1e0920f065bd14dd7/wandb/util.py#L482)) is being used anywhere at the moment, while the commented-out `to_json()` method below is defined for a bunch of different classes in `data_types.py`.

I think json_encode makes the most sense.   Lets prefix it with underscore "_json_encode" since Summary is a public object.

**Comment by vanpelt** [comment]
Yep, what @raubitsj said 👍

File: wandb/interface/interface.py [hunk 1]
```diff
@@ -214,14 +216,40 @@ def _summary_encode(self, value, path_from_root):
...
+        for item in summary_record.update:
+            pb_summary_item = pb_summary_record.update.add()
+            key_length = len(item.key)
+
+            assert key_length > 0
+
+            pb_summary_item.key = item.key[0]
+
+            if key_length > 1:
+                pb_summary_item.nested_key.extend(item.key[1:])
```
Comment (raubitsj) [inline wandb/interface/interface.py:231]: lets use nested_key anytime we are updating nested keys (key_length > 1) and leave key unset in that case

and key is set only if key_length == 1
I think that definition makes more sense than using the top level key and nest_keys for later keys.

**Pull request merged**